### PR TITLE
RAINCATCH-451: Update demo cloud to use raincatcher-store-simple module

### PR DESCRIPTION
**Pull request by JameelB**
## What
Update raincatcher-demo-cloud to use the new raincatcher-store-simple module https://github.com/feedhenry-raincatcher/raincatcher-simple-store

## Jira Ticket 
https://issues.jboss.org/browse/RAINCATCH-451

### CODE REVIEW

File: lib/app/file.js [hunk 1]
```diff
@@ -1,10 +1,15 @@
...
-var files = [];
+var SimpleStore = require('fh-wfm-simple-store');
 
 module.exports = function(mediator) {
-  var arrayStore = new ArrayStore('file', files);
-  arrayStore.listen('cloud:', mediator);
+  var fileStore = new SimpleStore('file');
+
+  fileStore.list().then(function(res) {
+    if (!res || res.length === 0) {
```
Comment (paolobueno) [inline lib/app/file.js]: I don't think you need initial `list` calls here. The fh.db implementation only touches the database upon `init(data)`, the constructor only initializes some local variables.

Comment (paolobueno) [inline lib/app/file.js]: Ah, now I see you're checking for existing data before initializing with the seed data, nvm!

Comment (nialldonnellyfh) [inline lib/app/file.js]: Can you add a comment here covering what is being checked and why. 

It should probably be noted that it should be removed in production applications as they are likely to be starting with 0 of everything.

File: lib/app/file.js [hunk 2]
```diff
@@ -1,10 +1,15 @@
...
+      fileStore.init();
+    }
+
+    fileStore.listen(':cloud', mediator);
```
Comment (nialldonnellyfh) [inline lib/app/file.js]: the `:cloud` is repeated several times. It would be nice to create a constant / config value for it.

File: lib/app/group.js [hunk 3]
```diff
@@ -20,9 +20,23 @@ var membership = [
 ];
 
 module.exports = function(mediator) {
-  var groupStore = new ArrayStore('group', groups);
-  groupStore.listen('cloud:', mediator);
+  var groupStore = new SimpleStore('group');
+  groupStore.list().then(function(res) {
```
Comment (nialldonnellyfh) [inline lib/app/group.js]: `res` is a bit generic. Can you rename these to be more descriptive (e.g. `groupList` )

File: lib/app/group.js [hunk 4]
```diff
@@ -20,9 +20,23 @@ var membership = [
...
-  var membershipStore = new ArrayStore('membership', membership);
-  membershipStore.listen('cloud:', mediator);
+    if (!res || res.length === 0) {
+      groupStore.init(groups);
+    }
+
+    groupStore.listen(':cloud', mediator);
+  });
+
+  var membershipStore = new SimpleStore('membership');
```
Comment (vchepeli) [inline lib/app/group.js]: there is duplication of functionality for each module. Create module store-init and then apply() on different modules

File: lib/app/result.js [hunk 5]
```diff
@@ -1,15 +1,7 @@
...
-  var resultsStore = new SimpleStore('result');
-
-  resultsStore.list().then(function(res) {
-    if (!res || res.length === 0) {
-      resultsStore.init();
-    }
-
-    resultsStore.listen(':cloud', mediator);
-  });
+  store.init('result', null, mediator);
```
Comment (paolobueno) [inline lib/app/result.js]: `null` here might break the in-memory implementation :sweat_smile:.

We can go ahead and test with it but `[]` is probably more meaningful.

File: lib/store-init.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    if (!storeData || storeData.length === 0) {
+      //Check if seed data is given, otherwise initialize store with no initial data.
+      if (seedData) {
+        dataStore.init(seedData);
+      } else {
+        dataStore.init();
+      }
+    }
+
+    dataStore.listen(config.get('topicPrefix'), mediator);
```
Comment (vchepeli) [inline lib/store-init.js]: `init()` returns Promise. You can still use `then(function(data))`
```js
then(function(result)){
console.log('dataStore initialized with data',data);
dataStore.listen(config.get('topicPrefix'), mediator);
```

**Comment by vchepeli** [comment]
LGTM

File: lib/store-init.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+ * @param {Object} mediator Object used for listening to a topic.
+ */
+function initStore(dataSetId, seedData, mediator) {
+  var dataStore = new SimpleStore(dataSetId);
+
+  /**
+   * Check if existing data exists. If no existing data is found, initialize with the given seed data. Otherwise,
+   * continue to listen to topic.
+   *
+   * TODO: Check for existing data may be removed in production as applications are likely to start with no
```
Comment (nialldonnellyfh) [inline lib/store-init.js]: Probably good to note that the seed data is only for demo purposes.

File: lib/store-init.js [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+   * initial data. Instead, initialize store with: dataStore.init();
+   */
+  dataStore.list().then(function(storeData) {
+    if (!storeData || storeData.length === 0) {
+      //Check if seed data is given, otherwise initialize store with no initial data.
+      //array store needs to be initialized with an empty array if there is no seed data given.
+      if (!seedData && process.env.WFM_USE_MEMORY_STORE === "true") {
+        seedData = [];
```
Comment (paolobueno) [inline lib/store-init.js]: We need to changet this check when we remove this env var from simple-store. Is not accepting `[]` still a bug in the $fh.db implementation?

Comment (JameelB) [inline lib/store-init.js]: @paolobueno Will be changing that check soon :) But yes, initializing fh-db store with [ ] is still giving me an error

Comment (paolobueno) [inline lib/store-init.js]: k, thanks, I'll go ahead and add fixes around these edge cases

File: Gruntfile.js [hunk 9]
```diff
@@ -138,15 +138,25 @@ module.exports = function(grunt) {
     },
     eslint: {
       src: ['*.js', 'lib/**/*.js', 'test/**/*.js']
+    },
+    mochaTest: {
+      test: {
+        options: {
+          run: true
+        },
+        src: ['lib/*spec.js']
```
Comment (vchepeli) [inline Gruntfile.js]: would write lib/**/*-spec.js

File: lib/app.js [hunk 10]
```diff
@@ -51,17 +52,20 @@ require('fh-wfm-user/lib/router/cloud')(mediator, app, authServiceGuid);
...
+module.exports = function() {
+  //make sure that all app modules finish setting up
+  return Promise.all([
+    require('./app/message')(mediator),
+    require('./app/workorder')(mediator),
+    require('./app/result')(mediator),
+    require('./app/workflow')(mediator),
+    require('./app/group')(mediator),
+    require('./app/file')(mediator)]).then(function() {
+      // Important that this is last!
```
Comment (vchepeli) [inline lib/app.js]: think it does not matter really, this promises still run in parallel

File: lib/config.js [hunk 11]
```diff
@@ -4,8 +4,10 @@ var cc = require('config-chain');
 
 var autoconfig = function(overrides) {
   var config = cc(overrides).add({
-    IP: process.env.OPENSHIFT_NODEJS_IP || '0.0.0.0'
-  , PORT: process.env.FH_PORT || process.env.OPENSHIFT_NODEJS_PORT || 8001
+    IP: process.env.OPENSHIFT_NODEJS_IP || '0.0.0.0',
+    PORT: process.env.FH_PORT || process.env.OPENSHIFT_NODEJS_PORT || 8001,
+    dataTopicPrefix: ':cloud:data',
+    persistentStore: (process.env.WFM_USE_MEMORY_STORE === "true")
```
Comment (vchepeli) [inline lib/config.js]: what parentheses useful here?

File: lib/app.js (hunk 10 above)
Comment (paolobueno) [inline lib/app.js]: This comment is actually about the error handler middleware below this line, see indentation.

Middleware chain setup is sensitive to the order they're give to the express app.

File: Gruntfile.js (hunk 9 above)
Comment (paolobueno) [inline Gruntfile.js]: :+1: the `**` is for globbing any amount of sub-directories, we already have subdirectories and would ideally be adding unit tests to those in the future.

File: lib/store-init-spec.js [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+};
+var SelectedStore = function(datasetId) {
+  this.datasetId = datasetId;
+};
+SelectedStore.prototype.list = function() {};
+SelectedStore.prototype.init = function() {};
+SelectedStore.prototype.listen = function() {};
+
+var mockSelectStore = function() {
+  return SelectedStore;
```
Comment (vchepeli) [inline lib/store-init-spec.js]: I'd return `new SelectedStore()` if you use prototype inheritance

**Comment by vchepeli** [comment]
LGTM

File: lib/store-init-spec.js (hunk 12 above)
Comment (JameelB) [inline lib/store-init-spec.js]: This SelectedStore needs to be returned as is, as initStore uses this to initialize data store: https://github.com/JameelB/raincatcher-demo-cloud/blob/a563e4b7ca49a6f0b6abdf39f1f9a4b42bf24a98/lib/store-init.js#L12

File: lib/app.js [hunk 13]
```diff
@@ -8,6 +8,7 @@ var mbaasApi = require('fh-mbaas-api')
   , cors = require('cors')
   , mediator = require('fh-wfm-mediator/lib/mediator')
   , bodyParser = require('body-parser')
+  , Promise = require('bluebird')
```
Comment (paolobueno) [inline lib/app.js]: Style: can you take the opportunity to change this commas to separate `var` statements? :D

File: lib/mocks/mockStore.js [hunk 14]
```diff
@@ -0,0 +1,49 @@
+var sinon = require('sinon');
```
Comment (paolobueno) [inline lib/mocks/mockStore.js]: IMO files dedicated to testing should still go under `/test/` and be required from the `-spec.js` files. Just so we don't have everything under `/lib`

@nialldonnellyfh WDYT?

File: lib/store-init-spec.js [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+    done();
+  });
+
+  it('should initialize store with given seed data', function(done) {
+    store.init('test-dataset-id', seedData, mediator).then(function() {
+      sinon.assert.callCount(mockList, 1);
+      sinon.assert.calledOnce(mockInit);
+      sinon.assert.calledWith(mockInit, seedData);
+      sinon.assert.calledOnce(mockListen);
+      done();
```
Comment (paolobueno) [inline lib/store-init-spec.js]: Should probably `return` the promises here instead of using the done callbacks, not sure if it catches thrown exceptions from the asserts or the test just times out.

File: lib/mocks/mockStore.js (hunk 14 above)
Comment (nialldonnellyfh) [inline lib/mocks/mockStore.js]: Yes, that makes sense. Any mocks / fixtures should be under the test folder.
